core/system/src/unvic.c

```diff
@@ -280,163 +282,169 @@ int unvic_irq_level_get(void)
-/* naked wrapper function needed to preserve LR */
-void UVISOR_NAKED unvic_svc_cx_in(uint32_t *svc_sp, uint32_t svc_pc)
+/** Performs a context switch-in as a result of an interrupt request.
+ *
+ * @internal
+ *
+ * This function is implemented as a wrapper, needed to make sure that the lr
+ * register doesn't get polluted and to provide context privacy during a context
+ * switch. The actual function is ::unvic_gateway_context_switch_in. The wrapper
+ * also changes the lr register so that we can return to a different privilege
+ * level. */
+void UVISOR_NAKED unvic_gateway_in(uint32_t svc_sp, uint32_t svc_pc)
 {
+    /* According to the ARM ABI, r0 and r1 will have the following values when
+     * this function is called:
+     *   r0 = svc_sp
+     *   r1 = svc_pc */
     asm volatile(
-        "push {r4 - r11}\n"         /* store callee-saved regs on MSP (priv) */
-        "push {lr}\n"               /* save lr for later */
-        // r0 = svc_sp              /* 1st arg: SVC sp */
-        // r1 = svc_pc              /* 1st arg: SVC pc */
-        "bl   __unvic_svc_cx_in\n"  /* execute handler and return */
-        "cmp  r0, #0\n"             /* check return flag */
-        "beq  no_regs_clearing\n"
-        "mov  r4,  #0\n"            /* clear callee-saved regs on request */
-        "mov  r5,  #0\n"
-        "mov  r6,  #0\n"
-        "mov  r7,  #0\n"
-        "mov  r8,  #0\n"
-        "mov  r9,  #0\n"
-        "mov  r10, #0\n"
-        "mov  r11, #0\n"
-        "no_regs_clearing:\n"
-        "pop  {lr}\n"               /* restore lr */
-        "orr  lr, #0x1C\n"          /* return with PSP, 8 words */
-        "bx   lr\n"
-        /* the unprivileged interrupt handler will be executed after this */
+        "push {r4 - r11}\n"                             /* Store the callee-saved registers on the MSP (privileged). */
+        "push {lr}\n"                                   /* Preserve the lr register. */
+        "bl   unvic_gateway_context_switch_in\n"        /* privacy = unvic_context_switch_in(svc_sp, svc_pc) */
+        "cmp  r0, #0\n"                                 /* if (privacy)  */
+        "beq  unvic_gateway_no_regs_clearing\n"         /* {             */
+        "mov  r4,  #0\n"                                /*     Clear r4  */
+        "mov  r5,  #0\n"                                /*     Clear r5  */
+        "mov  r6,  #0\n"                                /*     Clear r6  */
+        "mov  r7,  #0\n"                                /*     Clear r7  */
+        "mov  r8,  #0\n"                                /*     Clear r8  */
+        "mov  r9,  #0\n"                                /*     Clear r9  */
+        "mov  r10, #0\n"                                /*     Clear r10 */
+        "mov  r11, #0\n"                                /*     Clear r11 */
+        "unvic_gateway_no_regs_clearing:\n"             /* } else { ; }  */
+        "pop  {lr}\n"                                   /* Restore the lr register. */
```

Comment:
Here the approach is that the wrapper calls a dedicated handler (the actual switch-in/out functions) and then returns. If you are call-and-return-ing you always need to push the `lr` and then pop it again as the `pc`:

```
push {r4 - r11, lr}
... /* do some stuff */
pop {r4 - r11, pc}
```

We changed it slightly here (two consecutive `push` ops) because we will `pop` them separately: The first `pop` retrieves the `lr`, which is needed for exception return (in this case it's a `EXC_RETURN` `lr`). The second `pop` will be done at context-switch-out and will retrieve the callee-saved registers from the old stack.

The actual context switch functions have no expectations on their own stack frames, that's why we de-coupled the wrapper and the actual handler. In this way, the switch functions just follow the ABI, the wrappers do the dark magic.